test/common/websocket/codec_test.cc

```diff
@@ -0,0 +1,766 @@
+
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(1, frames.value().size());
+  EXPECT_EQ(2, frames.value()[0].opcode_);
+  EXPECT_TRUE(frames.value()[0].final_fragment_);
+  EXPECT_EQ(256, frames.value()[0].payload_length_);
+  EXPECT_EQ(false, frames.value()[0].masking_key_.has_value());
+}
+
+// 126 bytes binary message in a single masked frame
+TEST(WebSocketCodecTest, decode16BitBinaryMaskedFrame) {
+  Buffer::OwnedImpl buffer;
+  Buffer::addSeq(
+      buffer,
+      {
+          0x82, 0xfe, 0x00, 0x7e, 0xa1, 0xe8, 0xd7, 0xb0, 0x75, 0xe6, 0x2e, 0x56, 0x91, 0x66, 0xb4,
+          0x89, 0x66, 0xa3, 0xd4, 0xaf, 0x15, 0x3e, 0x7d, 0xa0, 0xd6, 0xca, 0xf9, 0xc7, 0xd9, 0x5d,
+          0xc6, 0x33, 0x9d, 0xd5, 0x41, 0x22, 0x0f, 0x18, 0xc1, 0xe7, 0xff, 0x8f, 0x2e, 0x6c, 0xaa,
+          0x81, 0x0f, 0x4e, 0xa5, 0xb6, 0x25, 0x4e, 0x8e, 0x3b, 0x25, 0x1e, 0x14, 0x58, 0x12, 0x5f,
+          0xfc, 0xe8, 0xb8, 0xc8, 0xc9, 0xd0, 0xa1, 0x53, 0x04, 0xdb, 0x91, 0xd6, 0x9a, 0xc6, 0x49,
+          0x6d, 0x96, 0x41, 0x8d, 0x9a, 0xd8, 0x00, 0x96, 0x41, 0x8d, 0x54, 0xdf, 0x0a, 0x87, 0xea,
+          0xe7, 0x4c, 0xdf, 0x16, 0x9a, 0x30, 0xa0, 0x26, 0x1c, 0xc4, 0x2c, 0xad, 0xfd, 0xd0, 0x2a,
+          0x16, 0xc5, 0xb4, 0xd7, 0x26, 0xa8, 0xca, 0x78, 0xe2, 0x1d, 0xe1, 0x8c, 0xde, 0xa8, 0x35,
+          0xd7, 0xda, 0x3a, 0x07, 0xb2, 0xf7, 0x71, 0x54, 0xd7, 0x26, 0xa8, 0xca, 0x78, 0xe2,
+      });
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(1, frames.value().size());
+  EXPECT_EQ(2, frames.value()[0].opcode_);
+  EXPECT_TRUE(frames.value()[0].final_fragment_);
+  EXPECT_EQ(126, frames.value()[0].payload_length_);
+  EXPECT_EQ(true, frames.value()[0].masking_key_.has_value());
+  EXPECT_EQ(0xa1e8d7b0, frames.value()[0].masking_key_.value());
+}
+
+// 64KiB binary message in a single unmasked frame
+// 0x82 0x7f 0x0000000000010000 [65536 bytes of binary data]
+TEST(WebSocketCodecTest, decode64BitBinaryUnmaskedFrame) {
+  Buffer::OwnedImpl buffer;
+  Buffer::addSeq(buffer, {0x82, 0x7F, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00});
+
+  // Add test payload 65536 bytes
+  for (uint16_t i = 0; i < 4096; ++i) {
+    Buffer::addSeq(buffer, {0x16, 0xc5, 0xb4, 0xd7, 0x26, 0xa8, 0xca, 0x78, 0xe2, 0x1d, 0xe1, 0x8c,
+                            0xde, 0xa8, 0x35, 0xde});
+  }
+
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(2, frames.value()[0].opcode_);
+  EXPECT_TRUE(frames.value()[0].final_fragment_);
+  EXPECT_EQ(65536, frames.value()[0].payload_length_);
+  EXPECT_EQ(false, frames.value()[0].masking_key_.has_value());
+}
+
+// Frame with only frame header data
+TEST(WebSocketCodecTest, decodeIncompleteFrameWithgOnlyFrameHeader) {
```

Comment:
I think it would be a good idea to have a test which does something like:
* Start with a complete frame in some array.
* Call decode 1 byte at a time to make sure we end up correctly parsing the frame at the end.